File: Data Structures & Algorithms/max-water-container/submission-2.java
class Solution {
    public int maxArea(int[] heights) {
        int n = heights.length;
        int left = 0;
        int right = n - 1;
        int max_area = 0;

        while (left < right){
            int area = Math.min(heights[left], heights[right]) * Math.abs(left - right);
            max_area = Math.max(max_area, area);

            if (heights[left] < heights[right]) {
                left++;
            } else {
                right--;
            }
        }
        
        return max_area;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0

        m = len(grid)
        n = len(grid[0])

        max_area = 0

        def dfs(x, y):
            if not (0 <= x < m) or not(0 <= y < n) or grid[x][y] != 1:
                return 0

            grid[x][y] = -1
            
            area = 1

            for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                nx, ny = dx + x, dy + y
                area += dfs(nx, ny)
            return area

        for i in range(m):
            for j in range(n):
                max_area = max(max_area, dfs(i, j))

        return max_area
